PHOENIX-538 Support UDFs

chrajeshbabu opened the pull request:
Patch to support UDFs. It mainly includes
- create temporary/permanent function query parsing
- storing function info
- dynamically loading udf jars.
- resolve functions 
- making use of udfs in different queries
- drop function.
- it tests

samarthjain commented on phoenix-core/src/main/java/org/apache/phoenix/cache/GlobalCache.java:
```diff
@@ -157,4 +159,18 @@ public TenantCache getChildTenantCache(ImmutableBytesWritable tenantId) {
+        }
+
+        @Override
+        public boolean equals(Object obj) {
```
Should you be overriding the hashCode() method too?

chrajeshbabu replied:
I will override the hashcode and just call super.hashCode() then we won't get checkstyle or findbug warning.

samarthjain commented on phoenix-core/src/main/java/org/apache/phoenix/compile/CreateFunctionCompiler.java:
```diff
@@ -0,0 +1,84 @@
+        final PhoenixConnection connection = statement.getConnection();
+        PhoenixConnection connectionToBe = connection;
+        Scan scan = new Scan();
+        final StatementContext context = new StatementContext(statement, FromCompiler.EMPTY_TABLE_RESOLVER, scan, new SequenceManager(statement));
```
Minor nit: You can use the other constructor here public StatementContext(PhoenixStatement statement)

chrajeshbabu replied:
That would be great. Will change.

samarthjain commented on phoenix-core/src/main/java/org/apache/phoenix/compile/FromCompiler.java:
```diff
@@ -383,6 +448,85 @@ protected TableRef createTableRef(NamedTableNode tableNode, boolean updateCacheI
+            return functions;
+        }
+
+        protected List<PFunction> createFunctionRef(List<String> functionNames, boolean updateCacheImmediately) throws SQLException {
```
Minor nit: is this function overriden somewhere? If not, it could be made private

chrajeshbabu replied:
no will change it to private.

samarthjain commented on phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:
```diff
@@ -700,6 +1007,29 @@ private PTable loadTable(RegionCoprocessorEnvironment env, byte[] key,
+            arrayList.add(key);
+            List<PFunction> functions = buildFunctions(arrayList, region, asOfTimeStamp);
+            if(functions != null) return functions.get(0);
+            // if not found then check if newer table already exists and add delete marker for timestamp
```
newer function?

chrajeshbabu replied:
yes

samarthjain commented on phoenix-core/src/main/java/org/apache/phoenix/expression/function/UDFExpression.java:
```diff
@@ -0,0 +1,217 @@
+        }
+        if (jarPath == null || jarPath.isEmpty() || config.get(DYNAMIC_JARS_DIR_KEY) != null
+                && (parent != null && parent.equals(config.get(DYNAMIC_JARS_DIR_KEY)))) {
+            Lock lock = locker.acquireLock(tenantId.getString());
```
Rajesh says: Just want to clarify one thing here the map is concurrent map so get and putifAbsent will be atomic but there is a chance like between get and put if other client already puts a class loader for the tenant then the classloader and configuration object created become garbage. That's why using locker. If that's ok I can do as you suggested. What do u say?

Samarth says: How expensive is creating a new classloader and config? To me it doesn't look too expensive. So I wouldn't worry about creating and discarding the two. Besides that race condition of another thread winning the race between get() and putIfAbsent is not very likely. So to sum it up, yes, I would get rid of the KeyLocker altogether.

chrajeshbabu replied:
Sure then will remove it.

samarthjain commented on phoenix-core/src/test/java/org/apache/phoenix/parse/QueryParserTest.java:
```diff
@@ -289,24 +289,6 @@ public void testNegativeCountStar() throws Exception {
     }
 
     @Test
-    public void testUnknownFunction() throws Exception {
```
Is there a reason why this test was removed? Are we no longer throwing an error? Or has the exception code changed? I would think if the user doesn't have load udf enabled then this error is still valid.

chrajeshbabu replied:
Now the exception will be thrown during compilation not while parsing because while parsing not able check whether udf enabled or not. That's why removed it.

samarthjain commented:
Almost there Rajesh. Just some minor nits and questions, otherwise LGTM.

chrajeshbabu commented:
Thanks Samarth for reviews. Will update patch addressing the comments.

JamesRTaylor commented on phoenix-core/src/main/java/org/apache/phoenix/expression/function/UDFExpression.java:
```diff
@@ -0,0 +1,217 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.MapMaker;
+
+public class UDFExpression extends ScalarFunction {
```
Having getDeterminism() return PER_INVOCATION is the only way to get each thread to have its own copy of the expression tree. Can  you file a JIRA with the failing test and exception? Is it only for functional indexing that you run into an issue?

chrajeshbabu replied:
if (expression.getDeterminism() != Determinism.ALWAYS) {
                        throw new SQLExceptionInfo.Builder(SQLExceptionCode.NON_DETERMINISTIC_EXPRESSION_NOT_ALLOWED_IN_INDEX).build().buildException();
                    }

Here it's the reason for failure James...getting problem with functional indexes only.

chrajeshbabu replied:
java.sql.SQLException: ERROR 521 (42898): Non-deterministic expression not allowed in an index
	at org.apache.phoenix.exception.SQLExceptionCode$Factory$1.newException(SQLExceptionCode.java:386)
	at org.apache.phoenix.exception.SQLExceptionInfo.buildException(SQLExceptionInfo.java:145)
	at org.apache.phoenix.schema.MetaDataClient.createIndex(MetaDataClient.java:1179)
	at org.apache.phoenix.compile.CreateIndexCompiler$1.execute(CreateIndexCompiler.java:95)
	at org.apache.phoenix.jdbc.PhoenixStatement$3.call(PhoenixStatement.java:303)
	at org.apache.phoenix.jdbc.PhoenixStatement$3.call(PhoenixStatement.java:1)
	at org.apache.phoenix.call.CallRunner.run(CallRunner.java:53)
	at org.apache.phoenix.jdbc.PhoenixStatement.executeMutation(PhoenixStatement.java:294)
	at org.apache.phoenix.jdbc.PhoenixStatement.execute(PhoenixStatement.java:1247)
	at org.apache.phoenix.end2end.UserDefinedFunctionsIT.testFunctionalIndexesWithUDFFunction(UserDefinedFunctionsIT.java:509)

JamesRTaylor replied:
I see - forgot about that, and we shouldn't change that. Instead, in the RowProjector constructor we'll want to set cloneRequired to true if any of the expressions use a UDF. The easiest way is probably to add a boolean hasUDFs to the RowProjector, add a ColumnResolver.hasUDFs() that returns true if !functions.isEmpty(). This will still clone if a UDF is used in the WHERE clause, but not in the SELECT expressions, but that's ok (it won't harm anything).

chrajeshbabu replied:
Added this James. Thanks for alternative soln.

JamesRTaylor commented on phoenix-core/src/main/antlr3/PhoenixSQL.g:
```diff
@@ -114,6 +114,15 @@ tokens
     UNION='union';
+    FUNCTION='function';
+    AS='as';
+    REPLACE='replace';
```
If support for REPLACE won't make this pull request (which is fine b/c it's already big enough), we should just remove it from the grammar for now.

chrajeshbabu replied:
Removed replace currently.

chrajeshbabu commented:
Thanks @JamesRTaylor  @samarthjain 
I have addressed the review comments and added to pull request. If it's ok I will commit this tomorrow morning IST and work on subtasks.

chrajeshbabu commented:
It's committed. Hence closing.
